lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py

```diff
@@ -339,102 +774,242 @@ def train_embedding_model(
+    optimizer_type: PyKEENSupportedOptimizers = PyKEENSupportedOptimizers.Adam,
+    epochs: int = 5,
+    training_approach: TrainingType = TrainingType.sLCWA,
+):
+    bundle = bundle.copy()
+    model_wrapper: PyKEENModelWrapper = bundle.other.get(model_name)
+    model = model_wrapper.model
+    transductive_training = prepare_triples(
+        bundle.dfs[transductive_table_name][["head", "relation", "tail"]],
+        relation_to_id=model_wrapper.relation_to_id,
+        inv_triples=True,
+    )
+    inductive_inference = prepare_triples(
+        bundle.dfs[inductive_inference_table][["head", "relation", "tail"]],
+        entity_to_id=model_wrapper.entity_to_id,
+        relation_to_id=model_wrapper.relation_to_id,
+        inv_triples=True,
+    )
+    inductive_validation = prepare_triples(
+        bundle.dfs[inductive_validation_table][["head", "relation", "tail"]],
+        entity_to_id=model_wrapper.entity_to_id,
+        relation_to_id=model_wrapper.relation_to_id,
+        inv_triples=True,
+    )
+    training_loop_cls = (
+        SLCWATrainingLoop if training_approach == TrainingType.sLCWA else LCWATrainingLoop
+    )
+    loop_kwargs = (
+        dict(
+            negative_sampler_kwargs=dict(num_negs_per_pos=32),
+        )
+        if training_approach == TrainingType.sLCWA
+        else dict()
+    )
+    training_loop = training_loop_cls(
+        triples_factory=transductive_training,
+        model=model,
+        optimizer=optimizer_type,
+        mode="training",
+        **loop_kwargs,
+    )
+
+    valid_evaluator = evaluation.SampledRankBasedEvaluator(
+        mode="validation",
+        evaluation_factory=inductive_validation,
+        additional_filter_triples=inductive_inference.mapped_triples,
+    )
+
+    early_stopper = stoppers.EarlyStopper(
+        model=model,
+        training_triples_factory=inductive_inference,
+        evaluation_triples_factory=inductive_validation,
+        frequency=5,
+        patience=40,
+        metric="ah@k",
+        result_tracker=None,
+        evaluation_batch_size=256,
+        evaluator=valid_evaluator,
+    )
+
+    losses = training_loop.train(
+        triples_factory=transductive_training,
+        stopper=early_stopper,
+        num_epochs=epochs,
+    )
+
+    model_wrapper.trained = True
+
+    bundle.dfs["training"] = pd.DataFrame({"training_loss": losses})
+    bundle.dfs["early_stopper_metric"] = pd.DataFrame(
+        {"early_stopper_metric": early_stopper.results}
+    )
+    bundle.other[model_name] = model_wrapper
+
+    return bundle
+
+
+@op("View early stopping metric", view="visualization")
+def view_early_stopping(bundle: core.Bundle):
+    metric = bundle.dfs["early_stopper_metric"].early_stopper_metric.tolist()
+    v = {
+        "title": {"text": "Early Stopping Metric"},
+        "xAxis": {"type": "category"},
+        "yAxis": {"type": "value"},
+        "series": [{"data": metric, "type": "line"}],
+    }
```

Comment:
Oh that's nice! First time I see a really _simple_ visualization with ECharts. 😊